runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientConnectionPool.java

```diff
@@ -1086,13 +1111,50 @@ private void onConnectionData(
             assert replyAck <= replySeq;
             assert replySeq <= replyAck + replyMax;
 
-            if ((flags & FLAG_INIT) != 0x00)
+            final DirectBuffer buffer = payload.buffer();
+            final int limit = payload.limit();
+            int progress = payload.offset();
+
+            while (progress < limit)
             {
-                long initialId = correlations.removeLong();
+                final int beforeResponseBytes = responseBytes;
+                if (responseBytes == 0)
+                {
+                    final ResponseHeaderFW responseHeader = responseHeaderRO.wrap(buffer, progress, limit);
+                    responseBytes = responseHeader.length() + KAFKA_FRAME_LENGTH_FIELD_OFFSET;
+                }
+
+                final int responseBytesMin = Math.min(responseBytes, payload.sizeof());
+                responseBytes -= responseBytesMin;
+                assert responseBytes >= 0;
+
+                long initialId = correlations.peekLong();
 
                 KafkaClientStream stream = streamsByInitialIds.get(initialId);
 
-                stream.doStreamData(traceId, flags, sequence, acknowledge, reserved, payload, extension);
+                int newFlags = flags;
+
+                if (beforeResponseBytes == 0 && responseBytes > 0)
+                {
+                    newFlags = FLAG_INIT;
+                }
+                else if (beforeResponseBytes != 0 && responseBytes == 0)
+                {
+                    newFlags = FLAG_FIN;
+                }
+                else if (responseBytes > 0)
+                {
+                    newFlags = FLAG_NONE;
+                }
```

Comment:
I think the flags calculation here is pretending the complete response is a single fragmented message spanning multiple DATA frames.

But instead, I think we want to preserve the flags sent by the response stream, with the exception of a minor adjustment when DATA frame does not belong to a single response.